server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -314,30 +281,106 @@ private boolean buildBufferAndMatchRight() {
     }
 
     private void addToBuffer(Object[] currentRow, int hash) {
-        Values existingRows = buffer.get(hash);
-        if (existingRows == null) {
-            existingRows = new Values();
-            buffer.put(hash, existingRows);
+        HashGroup hashGroup = buffer.get(hash);
+        if (hashGroup == null) {
+            hashGroup = new HashGroup(emitUnmatchedRows);
+            buffer.put(hash, hashGroup);
         }
-        existingRows.items.add(currentRow);
-        numberOfRowsInBuffer++;
+        hashGroup.add(currentRow);
+        numberOfHashGroupsInBuffer++;
     }
 
     private boolean findMatchingRows() {
-        while (leftMatchingRowsIterator.hasNext()) {
-            leftRow.cells(leftMatchingRowsIterator.next());
+        while (matchedHashGroupRowsIterator.hasNext()) {
+            int currentIdx = matchedHashGroupRowsIterator.nextIndex();
+            leftRow.cells(matchedHashGroupRowsIterator.next());
             combiner.setLeft(leftRow);
             if (joinCondition.test(combiner.currentElement())) {
+                matchedHashGroup.markMatched(currentIdx);
                 return true;
             }
         }
+        matchedHashGroup = null;
+        matchedHashGroupRowsIterator = null;
         return false;
     }
 
-    private static final class Values {
+    private final class HashGroup implements Iterable<Object[]> {
+
+        private final ArrayList<Object[]> rows;
+        @Nullable
+        private final BitSet rowIsJoinedFlags;
+
+        public HashGroup(boolean emitUnmatchedRows) {
+            circuitBreaker.addEstimateBytesAndMaybeBreak(
+                RamUsageEstimator.shallowSizeOfInstance(ArrayList.class),
+                "hash-group-rows-list"
+            );
+            this.rows = new ArrayList<>();
+
+            if (emitUnmatchedRows) {
+                circuitBreaker.addEstimateBytesAndMaybeBreak(
+                    RamUsageEstimator.shallowSizeOfInstance(BitSet.class) + RamUsageEstimator.NUM_BYTES_ARRAY_HEADER,
+                    "hash-group-bitset-initial"
+                );
+                this.rowIsJoinedFlags = new BitSet();
+            } else {
+                this.rowIsJoinedFlags = null;
+            }
+        }
+
+        public void add(Object[] row) {
+            rows.add(row);
+        }
+
+        public void markMatched(int idx) {
+            if (rowIsJoinedFlags != null) {
+                assert idx >= rowIsJoinedFlags.length()
+                    : "markMatched(idx) must be called with monotonically increasing idx; BitSet memory accounting depends on this invariant";
+                int prevNumberOfWords = 1; // pre-allocated 1 word during construction
+                int currentNumberOfWords = (idx + Long.SIZE - 1) >>> 6; // equivalent to ceil(idx+1 / 64)
+                circuitBreaker.addEstimateBytesAndMaybeBreak((currentNumberOfWords - prevNumberOfWords) * Long.BYTES, "hash-group-bitset-grow");
```

Comment:
`>>>6` is  `div 64` but `idx + Long.SIZE - 1` != `idx + 1`?
Maybe worth using prev size info to make it more readable?